src/control/security/grpc_cert_configs.go

```diff
@@ -0,0 +1,92 @@
+//
+// (C) Copyright 2020 Intel Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+// The Government's rights to use, modify, reproduce, release, perform, display,
+// or disclose this software are subject to the terms of the Apache License as
+// provided in Contract No. 8F-30005.
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+//
+
+// +build go1.15
+
+package security
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+
+	"github.com/pkg/errors"
+)
+
+func serverTLSConfig(cfg *TransportConfig) *tls.Config {
+	return &tls.Config{
+		ClientAuth:               tls.RequireAndVerifyClientCert,
+		Certificates:             []tls.Certificate{*cfg.tlsKeypair},
+		ClientCAs:                cfg.caPool,
+		MinVersion:               tls.VersionTLS12,
+		MaxVersion:               tls.VersionTLS12,
+		PreferServerCipherSuites: true,
+		CipherSuites: []uint16{
+			tls.TLS_RSA_WITH_AES_128_GCM_SHA256,
+			tls.TLS_RSA_WITH_AES_256_GCM_SHA384,
+		},
+		VerifyConnection: func(cs tls.ConnectionState) error {
+			opts := x509.VerifyOptions{
+				Roots:         cfg.caPool,
+				Intermediates: x509.NewCertPool(),
+				KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+			}
+			for _, cert := range cs.PeerCertificates[1:] {
+				opts.Intermediates.AddCert(cert)
+			}
+			_, err := cs.PeerCertificates[0].Verify(opts)
+			return err
+		},
+	}
+}
+
+func clientTLSConfig(cfg *TransportConfig) *tls.Config {
+	return &tls.Config{
+		Certificates:             []tls.Certificate{*cfg.tlsKeypair},
+		RootCAs:                  cfg.caPool,
+		MinVersion:               tls.VersionTLS12,
+		MaxVersion:               tls.VersionTLS12,
+		PreferServerCipherSuites: true,
+		CipherSuites: []uint16{
+			tls.TLS_RSA_WITH_AES_128_GCM_SHA256,
+			tls.TLS_RSA_WITH_AES_256_GCM_SHA384,
+		},
+		InsecureSkipVerify: true,
```

Comment:
It might be helpful to add some comments here explaining the choices made, for future reference. For example, noting that `InsecureSkipVerify: true` is required to enable custom verification. A casual reader might see that option and incorrectly assume that it means that we're not doing client-side verification.

Likewise, some comments explaining the logic in the VerifyConnection callback would be helpful for reviewers and future readers.